Migrate kubemark to e2e-up/e2e-down scripts

mborsz opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

It removes custom master setup logic from kubemark.

**Which issue(s) this PR fixes**:

https://github.com/kubernetes/kubernetes/issues/78726

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

mborsz commented:
/cc @wojtekt @mm4tt

mborsz commented:
/cc @wojtek-t

mborsz commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mborsz commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mborsz commented:
/retest

mborsz commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mborsz commented:
/retest

mborsz commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mborsz committed:
Migrate kubemark to e2e-up/e2e/down scripts.

wojtek-t commented on test/kubemark/gce/util.sh:
```diff
@@ -31,127 +31,32 @@ function authenticate-docker {
-# No. of retries is '$2' (if provided) or 1 (default).
-function execute-cmd-on-master-with-retries() {
-  RETRIES="${2:-1}" run-gcloud-compute-with-retries ssh "${MASTER_NAME}" --zone="${ZONE}" --project="${PROJECT}" --command="$1"
+function create-kubemark-master {
```
One thing to ensure (maybe it's already done) is that kubemark was restructured so that it can be run on different providers and IIRC it's now mostly about implementing this interface:
https://github.com/kubernetes/kubernetes/tree/master/test/kubemark/skeleton

Given people are running kubemarks on other providers, we don't want to break that.

mborsz replied:
The only logic for starting kubemark that lives in k8s repo is iks. This PR won't break it as iks/startup.sh copies all the script with 'exit 0' at the very end.

After changing the interface (IIUC) it should be to remove copy from iks/startup.sh (the new interface can be implemented for iks).

mborsz commented:
I think it's ready for review.

This PR contains also two other that are required for this one to work. I'm planning to submit them first:
* https://github.com/kubernetes/kubernetes/pull/78727: "Modify kube-up to support cluster without nodes."
* https://github.com/kubernetes/kubernetes/pull/78728: "Modify firewall rules names to make them shorter."

mborsz commented:
Rebased, I think it's ready for review

Adding hold to block merging until I run kubemark-5000 test with this and make sure it works.
/hold

mborsz commented:
/retest

mborsz commented:
/retest

wojtek-t commented on cluster/kubemark/gce/config-default.sh:
```diff
@@ -22,93 +22,20 @@
 # shellcheck disable=SC2034 # Variables sourced in other scripts.
 
 source "${KUBE_ROOT}/cluster/gce/config-common.sh"
+NUM_NODES="${KUBEMARK_NUM_NODES:-}"
```
What about:
NUM_WINDOWS_NODES
MASTER_SIZE
MASTER_DISK_SIZE
MASTER_ROOT_DISK_SIZE
?

Those are things that we allow overwriting differently for kubemark (similarly as we allow overwriting for regular gce clusters) and I think it's actually useful ability to have

wojtek-t replied:
Never mind - I see that is happening below.

wojtek-t commented on test/kubemark/gce/util.sh:
```diff
@@ -31,127 +31,54 @@ function authenticate-docker {
+    export KUBE_GCE_ENABLE_IP_ALIASES=true
+    export KUBE_GCE_NODE_IPAM_MODE=RangeAllocator
+    
+    # Disable all addons.
```
Can you add a comment that we enable (most of?) them by running dedicated pod in the base cluster?

mborsz replied:
Done.

wojtek-t commented on cluster/kubemark/gce/config-default.sh:
```diff
@@ -22,93 +22,20 @@
-RUNTIME_CONFIG="${KUBE_RUNTIME_CONFIG:-}"
-TERMINATED_POD_GC_THRESHOLD=${TERMINATED_POD_GC_THRESHOLD:-100}
-KUBE_APISERVER_REQUEST_TIMEOUT=300
-ETCD_COMPACTION_INTERVAL_SEC="${KUBEMARK_ETCD_COMPACTION_INTERVAL_SEC:-}"
```
I think we need this one (IIRC we're overwriting it in 5k-kubemark job).

mborsz replied:
I moved this (and all KUBEMARK_* envs) to start-master func.

wojtek-t commented on test/kubemark/start-kubemark.sh:
```diff
@@ -368,12 +94,12 @@ current-context: kubemark-context"
-    --from-literal=cluster_autoscaler.kubeconfig="${CLUSTER_AUTOSCALER_KUBECONFIG_CONTENTS}" \
-    --from-literal=npd.kubeconfig="${NPD_KUBECONFIG_CONTENTS}" \
-    --from-literal=dns.kubeconfig="${KUBE_DNS_KUBECONFIG_CONTENTS}"
+    --from-file=kubelet.kubeconfig="${LOCAL_KUBECONFIG}" \
```
Previously we had different (user, token) set for every component.
I think it has a value by itself, and we should try to mimic such behavior (given it's how it looks in real cluster).

I don't want to block this PR on it (because it's great on its own), but please file an issue and add a TODO to fix that.

mborsz replied:
Added comment.

wojtek-t commented on test/kubemark/start-kubemark.sh:
```diff
@@ -474,42 +200,24 @@ function wait-for-hollow-nodes-to-run-or-timeout {
-detect-project &> /dev/null
-find-release-tars
 
 # We need master IP to generate PKI and kubeconfig for cluster.
```
This comment is now obsolete, right?

mborsz replied:
Yes.

wojtek-t reviewed (commented):
This is great - I added a couple comments.

mborsz commented:
PTAL

mborsz commented:
/retest

wojtek-t commented on test/kubemark/gce/util.sh:
```diff
@@ -31,127 +31,54 @@ function authenticate-docker {
+    export KUBE_ENABLE_L7_LOADBALANCING="none"
+    
+    # Set kubemark-specific overrides.
+    export NUM_NODES="${KUBEMARK_NUM_NODES:-}"
```
and windows nodes

mborsz replied:
Migrated to the for loop which translates all KUBEMARK_X env variables to X (including windows nodes).

wojtek-t commented on test/kubemark/start-kubemark.sh:
```diff
@@ -367,13 +93,15 @@ current-context: kubemark-context"
 
   # Create secret for passing kubeconfigs to kubelet, kubeproxy and npd.
+  # It's bad that all component shares the same kubeconfig.
+  # TODO(mborsz): Migrate all components to separate credentials.
```
Can you file an issue for it so that we could better track it?

mborsz replied:
https://github.com/kubernetes/kubernetes/issues/79883

wojtek-t reviewed (commented):
just very minor stuff

mborsz commented on cluster/kubemark/gce/config-default.sh:
```diff
@@ -22,93 +22,20 @@
 # shellcheck disable=SC2034 # Variables sourced in other scripts.
 
 source "${KUBE_ROOT}/cluster/gce/config-common.sh"
+NUM_NODES="${KUBEMARK_NUM_NODES:-}"
```
In fact It's safe to delete this from config-default.sh as it's in util.sh file.

wojtek-t commented:
LGTM - though please ensure tests are passing (they may be unrelated though)

mborsz commented:
/retest

mborsz commented:
/retest

mborsz commented:
/retest

mborsz commented:
/retest

mborsz commented:
The presubmit failure was a valid one -- In one of the previous attempts, I removed NUM_NODES from config-default.sh, so start-kubemark was creating only 8 nodes instead of 500 required. This was blocking the test. I added the override back to config-default and I hope it should be fine now.

mborsz commented:
/retest

mborsz commented:
/retest

mborsz commented:
Presubmit passes now. Wojtek, could you take a look please?

wojtek-t commented:
This is great to see.

/lgtm
/approve

mborsz commented:
/hold cancel
I ran a big scale test with cluster created with this script and it looked fine.

wojtek-t commented:
/lgtm cancel

Given number of failure, I think something may still be wrong.

mborsz commented:
Yes, I checked few failures and the latency seems to be significantly higher than in the green ones. Will check that.

mborsz commented:
/retest

mborsz commented:
Yep, there was a bug in master sizing logic -- kubemark master was using MASTER_SIZE env variable that was set for 'root master' instead of recomputing it for 500 nodes.

wojtek-t commented on test/kubemark/gce/util.sh:
```diff
@@ -31,127 +31,65 @@ function authenticate-docker {
+    done
+
+    # Append kubemark-specific "MASTER_COMPONENTS_QPS_LIMITS".
+    # TODO(mborsz): Migrate all users of KUBEMARK_MASTER_COMPONENTS_QPS_LIMITS to {SCHEDULER,CONTROLLER_MANAGER}_TEST_ARGS.
```
Seems quite easy - it's used in exactly one place in our presets :)

wojtek-t reviewed (commented):
/lgtm
/approve
